syntax = "proto3";

package cyxwiz.protocol;

import "common.proto";
import "job.proto";

// Node information
message NodeInfo {
  string node_id = 1;
  string name = 2;
  Version version = 3;

  // Hardware
  repeated DeviceCaps devices = 4;
  int32 cpu_cores = 5;
  int64 ram_total = 6;
  int64 ram_available = 7;

  // Network
  string ip_address = 8;
  int32 port = 9;
  string region = 10;            // Geographic location

  // Performance
  double compute_score = 11;     // Benchmark score
  double reputation_score = 12;  // 0.0 to 1.0
  int64 total_jobs_completed = 13;
  int64 total_compute_hours = 14;
  double average_rating = 15;

  // Staking
  double staked_amount = 16;     // CYXWIZ tokens
  string wallet_address = 17;

  // Status
  bool is_online = 18;
  int64 last_heartbeat = 19;
  double uptime_percentage = 20;

  // Model deployment capabilities
  repeated string supported_formats = 21;  // e.g., ["ONNX", "GGUF", "PyTorch", "TensorFlow"]
  int64 max_model_size = 22;               // Maximum model size in bytes
  bool supports_terminal_access = 23;      // Can provide terminal/CLI access
  repeated string available_runtimes = 24; // e.g., ["onnxruntime", "pytorch", "llama.cpp"]

  // Pricing
  NodePricing pricing = 25;                // Pricing configuration
}

// Pricing configuration for compute nodes
message NodePricing {
  // Billing model type
  enum BillingModel {
    BILLING_HOURLY = 0;      // Charged per hour of compute time
    BILLING_PER_EPOCH = 1;   // Charged per training epoch
    BILLING_PER_JOB = 2;     // Fixed price per job
    BILLING_PER_INFERENCE = 3; // Charged per inference request (for deployment)
  }

  BillingModel billing_model = 1;

  // Base prices (in CYXWIZ tokens)
  double price_per_hour = 2;           // Price per hour of compute
  double price_per_epoch = 3;          // Price per training epoch
  double price_per_job_base = 4;       // Base price for per-job billing
  double price_per_inference = 5;      // Price per inference request

  // Minimum charges
  double minimum_charge = 6;           // Minimum charge per job (prevents micro-jobs)
  int32 minimum_duration_minutes = 7;  // Minimum billable duration

  // Discounts
  double discount_1h_plus = 8;         // Discount % for jobs > 1 hour (e.g., 0.05 = 5%)
  double discount_24h_plus = 9;        // Discount % for jobs > 24 hours
  double discount_bulk = 10;           // Discount % for bulk/repeat customers

  // Currency display (for UI)
  double usd_equivalent = 11;          // Current USD equivalent of price_per_hour
  int64 price_updated_at = 12;         // Unix timestamp of last price update

  // Accepted payment methods
  bool accepts_cyxwiz_token = 13;      // Accepts CYXWIZ token (default)
  bool accepts_sol = 14;               // Accepts SOL
  bool accepts_usdc = 15;              // Accepts USDC

  // Special pricing
  bool free_tier_available = 16;       // Offers free tier for testing
  int32 free_tier_minutes = 17;        // Minutes available in free tier
}

// Node registration
message RegisterNodeRequest {
  NodeInfo info = 1;
  string authentication_token = 2;
  bytes public_key = 3;           // For verification
}

message RegisterNodeResponse {
  StatusCode status = 1;
  string node_id = 2;
  string session_token = 3;
  Error error = 4;
}

// Node heartbeat
message HeartbeatRequest {
  string node_id = 1;
  NodeInfo current_status = 2;
  repeated string active_jobs = 3;
}

message HeartbeatResponse {
  StatusCode status = 1;
  bool keep_alive = 2;
  repeated string jobs_to_cancel = 3;
  Error error = 4;
}

// Job assignment (Central Server -> Server Node)
message AssignJobRequest {
  string node_id = 1;
  JobConfig job = 2;
  string authorization_token = 3;
}

message AssignJobResponse {
  StatusCode status = 1;
  string job_id = 2;
  bool accepted = 3;
  Error error = 4;
  string estimated_start_time = 5;
}

// Job progress report (Server Node -> Central Server)
message ReportProgressRequest {
  string node_id = 1;
  string job_id = 2;
  JobStatus status = 3;
  map<string, double> current_metrics = 4;
}

message ReportProgressResponse {
  StatusCode status = 1;
  bool continue_job = 2;         // Server can request cancellation
  Error error = 3;
}

// Job completion report
message ReportCompletionRequest {
  string node_id = 1;
  string job_id = 2;
  JobResult result = 3;
  bytes signature = 4;           // Node signature for verification
}

message ReportCompletionResponse {
  StatusCode status = 1;
  bool payment_released = 2;
  string payment_tx_hash = 3;
  Error error = 4;
}

// Node metrics query
message GetNodeMetricsRequest {
  string node_id = 1;
  int64 start_time = 2;
  int64 end_time = 3;
}

message GetNodeMetricsResponse {
  repeated MetricPoint metrics = 1;
  Error error = 2;
}

message MetricPoint {
  int64 timestamp = 1;
  double cpu_usage = 2;
  double gpu_usage = 3;
  double memory_usage = 4;
  double network_throughput = 5;
  double power_consumption = 6;
}

// P2P: Job acceptance notification for direct Engine→Node connection
message JobAcceptedRequest {
  string node_id = 1;
  string job_id = 2;
  string engine_address = 3;       // Where Engine connected from
  int64 accepted_at = 4;           // Unix timestamp
  string node_endpoint = 5;        // The endpoint Engine connected to
}

message JobAcceptedResponse {
  StatusCode status = 1;
  bool acknowledged = 2;
  Error error = 3;
}

// Node Service - Central Server to Server Node
service NodeService {
  // Node registration
  rpc RegisterNode(RegisterNodeRequest) returns (RegisterNodeResponse);

  // Keep-alive heartbeat
  rpc Heartbeat(HeartbeatRequest) returns (HeartbeatResponse);

  // Assign job to node
  rpc AssignJob(AssignJobRequest) returns (AssignJobResponse);

  // Report job progress
  rpc ReportProgress(ReportProgressRequest) returns (ReportProgressResponse);

  // Report job completion
  rpc ReportCompletion(ReportCompletionRequest) returns (ReportCompletionResponse);

  // Get node metrics
  rpc GetNodeMetrics(GetNodeMetricsRequest) returns (GetNodeMetricsResponse);

  // P2P: Notify that job was accepted directly from Engine
  rpc NotifyJobAccepted(JobAcceptedRequest) returns (JobAcceptedResponse);
}

// Node Discovery Service - For finding available nodes
service NodeDiscoveryService {
  // Find available nodes matching requirements
  rpc FindNodes(FindNodesRequest) returns (FindNodesResponse);

  // List all nodes
  rpc ListNodes(ListNodesRequest) returns (ListNodesResponse);

  // Get specific node info
  rpc GetNodeInfo(GetNodeInfoRequest) returns (GetNodeInfoResponse);
}

message FindNodesRequest {
  // Hardware requirements
  DeviceType required_device = 1;
  int64 min_memory = 2;                    // Minimum GPU/RAM in bytes
  int64 min_vram = 10;                     // Minimum VRAM in bytes (for GPU jobs)
  double min_compute_score = 11;           // Minimum benchmark score

  // Trust requirements
  double min_reputation = 3;
  double min_stake = 12;                   // Minimum staked CYXWIZ tokens

  // Location preferences
  string preferred_region = 4;
  repeated string allowed_regions = 13;    // Only consider these regions
  repeated string blocked_regions = 14;    // Exclude these regions

  // Pricing constraints
  double max_price_per_hour = 6;           // Maximum price in CYXWIZ tokens
  double max_price_usd = 7;                // Maximum price in USD (for convenience)
  NodePricing.BillingModel preferred_billing = 8;  // Preferred billing model
  bool require_free_tier = 9;              // Only nodes with free tier

  // Results
  int32 max_results = 5;

  // Sorting preference
  enum SortBy {
    SORT_BY_PRICE = 0;           // Cheapest first
    SORT_BY_PERFORMANCE = 1;     // Highest compute_score first
    SORT_BY_REPUTATION = 2;      // Highest reputation first
    SORT_BY_AVAILABILITY = 3;    // Most available (lowest queue) first
  }
  SortBy sort_by = 15;
}

message FindNodesResponse {
  repeated NodeInfo nodes = 1;
  int32 total_matching = 2;          // Total nodes matching criteria (for pagination info)
  double avg_price_per_hour = 3;     // Average price of matching nodes
  double min_price_per_hour = 4;     // Cheapest matching node
  double max_price_per_hour = 5;     // Most expensive matching node
  Error error = 6;
}

message ListNodesRequest {
  // Pagination
  int32 page_size = 1;
  string page_token = 2;

  // Filters
  bool online_only = 3;
  DeviceType filter_device = 4;        // Filter by device type
  string filter_region = 5;            // Filter by region

  // Price range filter (for marketplace browsing)
  double min_price_per_hour = 6;
  double max_price_per_hour = 7;

  // Sorting (reuse FindNodesRequest.SortBy enum)
  FindNodesRequest.SortBy sort_by = 8;
  bool sort_descending = 9;            // Default is ascending
}

message ListNodesResponse {
  repeated NodeInfo nodes = 1;
  string next_page_token = 2;
  int32 total_count = 3;

  // Marketplace statistics (for UI display)
  int32 online_count = 4;              // Number of currently online nodes
  double network_avg_price = 5;        // Average price across all nodes
  double network_total_compute = 6;    // Total compute power available (sum of compute_scores)
}

message GetNodeInfoRequest {
  string node_id = 1;
}

message GetNodeInfoResponse {
  NodeInfo info = 1;
  Error error = 2;
}

// ============================================================================
// Job Status Reporting Service (Server Node → Central Server)
// ============================================================================

// Update job status (progress or state change)
message UpdateJobStatusRequest {
  string node_id = 1;
  string job_id = 2;
  StatusCode status = 3;           // running, completed, failed
  double progress = 4;              // 0.0 to 1.0
  map<string, double> metrics = 5;  // loss, accuracy, etc.
  int32 current_epoch = 6;
  string log_message = 7;           // Optional progress message
}

message UpdateJobStatusResponse {
  StatusCode status = 1;
  bool should_continue = 2;         // Central Server can request cancellation
  Error error = 3;
}

// Report final job completion with results
message ReportJobResultRequest {
  string node_id = 1;
  string job_id = 2;
  StatusCode final_status = 3;     // success or failed

  // Final metrics
  map<string, double> final_metrics = 4;

  // Model output (if applicable)
  string model_weights_uri = 5;    // IPFS hash or storage location
  string model_weights_hash = 6;   // SHA256 for verification
  int64 model_size = 7;            // bytes

  // Compute stats
  int64 total_compute_time = 8;    // milliseconds
  string error_message = 9;        // If failed
}

message ReportJobResultResponse {
  StatusCode status = 1;
  Error error = 2;
}

// Job Status Service - Implemented by Central Server, called by Server Nodes
service JobStatusService {
  // Update job progress during execution
  rpc UpdateJobStatus(UpdateJobStatusRequest) returns (UpdateJobStatusResponse);

  // Report final job result (completion or failure)
  rpc ReportJobResult(ReportJobResultRequest) returns (ReportJobResultResponse);
}
